From: Jan Pokorny <jpokorny@redhat.com>
Date: Wed, 16 Nov 2011 21:40:07 +0200
Subject: [PATCH 4/6] fix bz742431: read all available with peer's receive()

The preprocessor conditional is kept for easy switch when needed.

Signed-off-by: Jan Pokorny <jpokorny@redhat.com>

diff -ur a/ricci/modules/cluster/clumon/src/daemon/Peer.cpp b/ricci/modules/cluster/clumon/src/daemon/Peer.cpp
--- a/ricci/modules/cluster/clumon/src/daemon/Peer.cpp	2011-11-16 21:16:29.402789607 +0100
+++ b/ricci/modules/cluster/clumon/src/daemon/Peer.cpp	2011-11-16 21:37:18.392337611 +0100
@@ -90,7 +90,18 @@
 	log("receiving data from " + _hostname, LogTransfer);
 
 	String& in = *_in;
-	in += _sock->recv();
+
+#define ENABLE_MULTIRECV 1
+#if ENABLE_MULTIRECV
+	int recv_len;
+	do {
+#endif
+		String received(_sock->recv());
+		in += received;
+#if ENABLE_MULTIRECV
+		recv_len = received.length();
+	} while (recv_len == SOCKET_RECV_BUFFER);
+#endif
 
 	vector<String> ret;
 
diff -ur a/ricci/modules/cluster/clumon/src/daemon/Peer.h b/ricci/modules/cluster/clumon/src/daemon/Peer.h
--- a/ricci/modules/cluster/clumon/src/daemon/Peer.h	2011-11-16 21:17:44.888723521 +0100
+++ b/ricci/modules/cluster/clumon/src/daemon/Peer.h	2011-11-16 21:40:24.858639311 +0100
@@ -48,6 +48,13 @@
 		*/
 		void send();
 
+		/*
+		** Try to receive available input, return array of particular messages.
+		**
+		** If not enough data received to form a whole XML status update,
+		** the already received part is kept back in _in which is then used
+		** upon next call as the prefix (empty array can be returned).
+		*/
 		std::vector<String> receive();
 
 		bool outq_empty() { return _out->empty(); }
